src/gasman.c

```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
 }
 
 
-/****************************************************************************
-**
-*F  SwapMasterPoint( <bag1>, <bag2> ) . . . swap pointer of <bag1> and <bag2>
-*/
-void SwapMasterPoint (
-    Bag                 bag1,
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are garbage collected, so we always put
+// *both* bags on the ChangedBags linked-list, rather than pick through the
+// exact cases, as it is never incorrect to mark something changed.
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// As ChangedBags is a singly linked list, we can't easily take anything out of
+// this list. Also, as the link pointers are stored inside the bag itself, when
+// we swap PTR_BAG, we must swap LINK_BAG as well.
+//
+// If both bags are old, and the original bag1 was changed, we need to now mark
+// bag2 changed (and vice-versa).
+//
+// If both bags are young, GAP assumes they are both changed.
+//
+// If bag1 was young and bag2 was old, then after swapping, bag1 is old and bag2
+// is young, as the 'young'ness moves with the contents, so we must mark bag1
+// changed if bag2 was previously changed.
+//
+// More importantly, bag2 is now young, but the only references to bag2 might be
+// in an old, unchanged bag, so it would get (incorrectly) garbage collected,
+// because these bags are not considered in a garbage collection.
+//
+// We therefore force bag2 onto the ChangedBags list, but we can't use
```

Comment:
"We therefore" makes it sound as if this is an obvious consequence, but to me it isn't, resp. to understand it I actually need the *next* paragraph. So how about this:

"To avoid this, we for bag2 onto the ChangedBags list (note that we cannot use CHANGED_BAG for this, as it skips young bags). This prevents bag2 from being collected, as GASMAN never collects any bags on the ChangedBag list."

Then the next paragraph can be removed, too.